.github/workflows/pytest.yml

```diff
@@ -0,0 +1,402 @@
+      restored: ${{ steps.restore-test-durations.outputs.cache-hit == '' && 'false' || 'true' }}
+    steps:
+      # https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow#example-using-an-output-to-define-two-matrice
+      - name: Define matrix split
+        id: splits
+        # default maxtrix for splits if we did not get an input
+        run: |
+          set -ex
+          export SPLIT_COUNT=${{ inputs.splits || 12 }}
+          if [ "$SPLIT_COUNT" -gt 24 ]; then
+            echo "Splits cannot exceed 24"
+            export SPLIT_COUNT=24
+          fi
+
+          if [ "$ACT" == "true" ] && [ -z "${{ inputs.splits }}" ]; then
+            echo "Running with act limiting splits to 1 as splits not provided"
+            export SPLIT_COUNT=1
+          else
+            echo "Running on GitHub Actions"
+          fi
+
+          SPLIT_LIST="["
+          for ((i=1; i<=SPLIT_COUNT; i++)); do
+            SPLIT_LIST+="\"$i\""
+            if [ $i -lt $SPLIT_COUNT ]; then
+              SPLIT_LIST+=", "
+            fi
+          done
+          SPLIT_LIST+="]"
+
+          echo "Generated list: $SPLIT_LIST"
+
+          export SPLIT_COUNT="$(echo ${SPLIT_LIST} | jq 'length')"
+          echo "${SPLIT_COUNT} : ${SPLIT_LIST}"
+          echo "SPLIT_LIST=${SPLIT_LIST}" >> "$GITHUB_OUTPUT"
+          echo "SPLIT_COUNT=${SPLIT_COUNT}" >> "$GITHUB_OUTPUT"
+
+      # It's mandatory to use the exact same path when saving/restoring cache, otherwise it won't work
+      # (the same key is not enough - see documentation:
+      # https://github.com/actions/cache/blob/main/README.md#cache-version).
+      # I went with `/tmp/.test_durations`.
+      - name: Restore test durations
+        id: restore-test-durations
+        uses: actions/cache/restore@v4
+        with:
+          path: /tmp/.test_durations
+          key: tests-durations-${{ github.sha }}
+          restore-keys: |
+            tests-durations-${{ github.sha }}-
+            tests-durations-
+          fail-on-cache-miss: false
+
+      # Then we upload the restored test durations as an artifact. This way, each matrix job will download
+      # it when it starts. When a matrix job will be manually retried, it will also reuse the artifact (to
+      # retry the exact same tests, even if the cache has been updated in the meantime).
+      - name: Upload test durations
+        if: steps.restore-test-durations.outputs.cache-hit != ''
+        uses: actions/upload-artifact@v4
+        with:
+          name: test-durations-before
+          path: /tmp/.test_durations
+          include-hidden-files: true
+
+  pytest:
+    needs: define-matrix
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        split: ${{ fromJSON(needs.define-matrix.outputs.SPLIT_LIST) }}
+
+    services:
+      ckan_postgres:
+        image: postgres:12 # should be postgres:15
```

Comment:
I've left this as what circleci users, but it should be on a lts supported version. 
12 is already end of support Nov 2024 
We should be on 13 (only gives us to Nov 2025). Postgres has a 5 year support window per Major Version.

https://www.postgresql.org/support/versioning/